src/go/web/handlers.go

```diff
@@ -3332,6 +3338,214 @@ func GetError(w http.ResponseWriter, r *http.Request) {
+	fileInfo, err := os.Stat(combinedPath)
+
+	if err != nil {
+		errString := fmt.Sprintf("Error getting path %s: %s", combinedPath, err.Error())
+		log.Error(errString)
+		http.Error(w, errString, http.StatusInternalServerError)
+		return
+	}
+	if !strings.HasPrefix(combinedPath, basePath) {
+		errString := fmt.Sprintf("Error getting path %s: Path is not within mount", combinedPath)
+		log.Error(errString)
+		http.Error(w, errString, http.StatusBadRequest)
+		return
+	}
+	if !fileInfo.IsDir() {
+		http.Error(w, fmt.Sprintf("Expected directory not file: %s", combinedPath), http.StatusBadRequest)
+		return
+	}
+
+	dirFiles, err := ioutil.ReadDir(combinedPath)
+	if err != nil {
+		errString := fmt.Sprintf("Error getting files in %s: %s", combinedPath, err.Error())
+		log.Error(errString)
+		http.Error(w, errString, http.StatusInternalServerError)
+	}
+
+	var files file.Files
+	for _, f := range dirFiles {
+		file := file.MakeFile(f, combinedPath)
+		file.Path = strings.TrimPrefix(file.Path, basePath)
+		files = append(files, file)
+	}
+	body, _ := json.Marshal(util.WithRoot("files", files))
+	w.Write(body)
+}
+
+// GET /experiments/{exp}/vms/{name}/files/download?path=
+func DownloadMountFile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	basePath := mm.GetLocalMountPath(vars["exp"], vars["name"])
+
+	role := r.Context().Value("role").(rbac.Role)
+	if !role.Allowed("vms/mount", "patch") {
+		http.Error(w, "forbidden", http.StatusForbidden)
+		return
+	}
+
+	mountsMu.RLock()
+	defer mountsMu.RUnlock()
+	_, exists := activeMounts[basePath]
+
+	if !exists {
+		http.Error(w, fmt.Sprintf("No existing mount for %s", basePath), http.StatusBadRequest)
+		return
+	}
+
+	combinedPath := filepath.Join(basePath, vars["path"])
+	log.Info("combinedPath: %s", combinedPath)
+	fileInfo, err := os.Stat(combinedPath)
+
+	if err != nil {
+		errString := fmt.Sprintf("Error getting path %s: %s", combinedPath, err.Error())
+		log.Error(errString)
+		http.Error(w, errString, http.StatusInternalServerError)
+		return
+	}
+	if !strings.HasPrefix(combinedPath, basePath) {
+		errString := fmt.Sprintf("Error getting path %s: Path is not within mount", combinedPath)
+		log.Error(errString)
+		http.Error(w, errString, http.StatusBadRequest)
+		return
+	}
+	if fileInfo.IsDir() {
+		http.Error(w, fmt.Sprintf("Can't download directory: %s", combinedPath), http.StatusBadRequest)
+		return
+	}
+	log.Info("Download for file: %s", fileInfo.Name())
+
+	w.Header().Set("Content-Disposition", "attachment; filename="+strconv.Quote(fileInfo.Name()))
+	w.Header().Set("Content-Type", "application/octet-stream")
+	http.ServeFile(w, r, combinedPath)
+}
+
+// PUT /experiments/{exp}/vms/{name}/files/download?path=
+func UploadMountFile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	basePath := mm.GetLocalMountPath(vars["exp"], vars["name"])
+
+	role := r.Context().Value("role").(rbac.Role)
+	if !role.Allowed("vms/mount", "patch") {
```

Comment:
Should pass `{exp_name}/{vm_name}` to `role.Allowed` to support per-VM access control for mounts.